sympy/matrices/normalforms.py

```diff
@@ -0,0 +1,136 @@
+from __future__ import print_function, division
+from sympy.matrices import diag
+
+'''
+Functions returning normal forms of matrices
+
+'''
+def smith_normal_form(m, domain = None):
+    '''
+    Return the Smith Normal Form of a matrix `m` over the ring `domain`.
+    This will only work if the ring is a principal ideal domain.
+
+    Examples
+    ========
+
+    >>> from sympy.polys.solvers import RawMatrix as Matrix
+    >>> from sympy.polys.domains import ZZ
+    >>> from sympy.matrices.normalforms import smith_normal_form
+    >>> m = Matrix([[12, 6, 4], [3, 9, 6], [2, 16, 14]])
+    >>> setattr(m, "ring", ZZ)
+    >>> print(smith_normal_form(m))
+    Matrix([[1, 0, 0], [0, 10, 0], [0, 0, -30]])
+
+    '''
+    invs = smith_normal_invariants(m)
+    smf = diag(*invs)
+    n = len(invs)
+    if m.rows > n:
+        smf = smf.row_insert(m.rows, zeros(m.rows-n, m.cols))
+    elif m.cols > n:
+        smf = smf.col_insert(m.cols, zeros(m.rows, m.cols-n))
+    return smf
+
+def smith_normal_invariants(m, domain = None):
+    '''
+    Return the tuple of abelian invariants for a matrix `m`
+    (as in the Smith-Normal form)
+
+    '''
+    if not domain:
+        if not (hasattr(m, "ring") and m.ring.is_PID):
+            raise ValueError(
+                "The matrix entries must be over a principal ideal domain")
+        else:
+            domain = m.ring
+
```

Comment:
We should check here that the matrix is not empty. If it is (i.e., `len(m) == 0`), then the empty list is returned.